bin/create_testbench_template.py

```diff
@@ -0,0 +1,181 @@
+import cocotb
+from cocotb.triggers import Timer
+from cocotb.result import TestError, TestFailure
+from cocotb.handle import *
+
+import logging
+import pprint
+
+# NOTE: port discovery needs SINGLETON_HANDLES in lib/gpi/Makefile not set!
+
+@cocotb.test()
+def create_testbench_template(dut):
+    """Generate a testbench template for the DUT."""
+
+    yield Timer(0)  # needed to make this a valid coroutine
+
+    input_ports_=[]
+    output_ports_=[]
+    inout_ports_=[]
+    supply_ports_=[]
+    ground_ports_=[]
+    clock_ports_=[]
+    reset_ports_=[]
+    for thing_ in dut:
+        if isinstance(thing_, HierarchyObject) and thing_._is_port:
+            dut._log.info('%s Port %s' % (thing_._port_direction_string[4:].capitalize(), thing_._path))
+            if thing_._port_direction_string == 'GPI_INPUT':
+                input_ports_.append(thing_)
+            elif thing_._port_direction_string == 'GPI_OUTPUT':
+                output_ports_.append(thing_)
+            elif thing_._port_direction_string == 'GPI_INOUT':
+                inout_ports_.append(thing_)
+            else:
+                dut._log.info('Unhandled port direction %s of %s!' % (thing_._port_direction_string, thing_._path))
+
+    # match names to identify "special" ports:
+    for port_ in set(input_ports_+inout_ports_):
+        if port_._name.startswith('vd') or port_._name.startswith('vc'):
+            # Examples: vdd, vddd, vdda, vdio, vcc, vcore
+            dut._log.info('Identified a supply port: %s' % (port_._name))
+            supply_ports_.append(port_)
+        elif port_._name.startswith('vs') or port_._name.startswith('gnd') or port_._name.endswith('gnd'):
+            # Examples: gnd, agnd, dgnd, vs, vss, vssd, vsio
+            dut._log.info('Identified a ground port: %s' % (port_._name))
+            ground_ports_.append(port_)
+        elif (port_._name.startswith('clk') or port_._name.endswith('clk')
+              or port_._name.startswith('clock') or port_._name.endswith('clock')):
+            # Examples: clk_2MHz, clk_main, main_clk, main_clock
+            dut._log.info('Identified a clock: %s' % (port_._name))
+            clock_ports_.append(port_)
+        elif port_._name in ('reset', 'resetn', 'reset_n', 'rst', 'rstn', 'rst_n'):
+            # Examples: exactly these
+            dut._log.info('Identified a reset: %s' % (port_._name))
+            reset_ports_.append(port_)
+
+    # clean up original lists:
+    for port_ in supply_ports_+ground_ports_+clock_ports_+reset_ports_:
+        try:
+            input_ports_.remove(port_)
+        except:
+            pass
+        try:
+            inout_ports_.remove(port_)
+        except:
+            pass
+
+    filename_ = os.getenv('COCOTB_TESTBENCH_TEMPLATE')
+    if filename_ is None:
+        filename_ = 'testbench_template'
+
+    with open(filename_+'.py', 'w') as f:
+        f.write("""# This is a generated testbench template for cocotb (https://cocotb.readthedocs.io)
```

Comment:
This is a good usecase for `textwrap.dedent`

Reply:
Good catch, done.